### PR TITLE
feat(notifier): disallow duplicate url registration

### PR DESCRIPTION
**Pull request by vjgene**
Fixes https://rsklabs.atlassian.net/browse/RMKT-779

### CODE REVIEW

File: src/components/organisms/notifier/provider/ProviderRegistrarForm.tsx [hunk 1]
```diff
@@ -101,3 +114,7 @@ const ProviderRegistrarForm: FC<ProviderRegistrarFormProps> = ({
 }
 
 export default ProviderRegistrarForm
+
+/* function AppContext<T>(AppContext: any): { state: { apis: { 'notifier/v0/providers': any } } } {
+  throw new Error('Function not implemented.')
+} */
```
Comment (itofarina) [inline src/components/organisms/notifier/provider/ProviderRegistrarForm.tsx]: Remove comment please 😄

**Comment by vjgene** [comment]
https://user-images.githubusercontent.com/11134695/123810169-7dd92480-d8f2-11eb-93b2-e878d3b7647d.mp4

File: src/components/organisms/notifier/provider/ProviderRegistrarForm.tsx [hunk 2]
```diff
@@ -27,11 +29,22 @@ const ProviderRegistrarForm: FC<ProviderRegistrarFormProps> = ({
...
+  const {
+    state: {
+      apis: {
+        [notifierProvidersAddress]: providersApi,
+      },
+    },
+  } = useContext<AppContextProps>(AppContext)
 
-  const validateProviderURL = (url: string): Promise<boolean> => {
+  const validateProviderURL = async (url: string): Promise<boolean> => {
```
Comment (itofarina) [inline src/components/organisms/notifier/provider/ProviderRegistrarForm.tsx]: Now that we have 3 different cases where the field could have an error, I wonder if we should as well update our error message to be more accurate. 
([reference](https://react-hook-form.com/api/useform/seterror))
We can open an issue if you think it should be done in another PR. What do you think?

**Review by itofarina — APPROVED** [review]
Looks good!
Just minor comments

File: src/api/rif-marketplace-cache/notifier/providers/api.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (result.length) {
+      this.errorReporter({
+        id: 'service-fetch',
+        text: 'The URL is already registered',
+        error: new Error('URL is already registered'),
+      })
+    }
+
+    return !result.length
```
Comment (jurajpiar) [inline src/api/rif-marketplace-cache/notifier/providers/api.ts]: consider:
```
  isRegisteredURL = async (url: string): Promise<boolean> => {
    const {length: hasResult} = await this.fetch({ url })

    if (!hasResult) {
      this.errorReporter({
        id: 'service-fetch',
        text: 'The URL is already registered',
        error: new Error('URL is already registered'),
      })
    }

    return hasResult
```

File: src/api/rif-marketplace-cache/notifier/providers/api.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    filters: NotifierProvidersFilters,
+  ): Promise<Provider[]> => {
+    const result: Paginated<Provider> = await this.service.find({
+      query: filters && {
+        ...filters,
+      },
+    })
+
+    const { data, ...metadata } = isResultPaginated(result)
+      ? result : { data: result }
```
Comment (jurajpiar) [inline src/api/rif-marketplace-cache/notifier/providers/api.ts]: You don't need this as you know you will never have paginated results.

File: src/api/rif-marketplace-cache/notifier/providers/api.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  _channel = wsChannel
+
+  constructor() {
+    super(client)
+  }
+
+  _fetch = async (
+    filters: NotifierProvidersFilters,
+  ): Promise<Provider[]> => {
+    const result: Paginated<Provider> = await this.service.find({
```
Comment (jurajpiar) [inline src/api/rif-marketplace-cache/notifier/providers/api.ts]: Not a paginated response here

File: src/api/rif-marketplace-cache/notifier/subscriptions/api.ts [hunk 6]
```diff
@@ -22,6 +22,7 @@ export const mapFromTransport = ({
   expirationDate,
   price,
   rateId,
+  provider,
```
Comment (jurajpiar) [inline src/api/rif-marketplace-cache/notifier/subscriptions/api.ts]:
```suggestion
  provider: { url },
```

File: src/api/rif-marketplace-cache/notifier/subscriptions/api.ts [hunk 7]
```diff
@@ -30,7 +31,7 @@ export const mapFromTransport = ({
   expirationDate: new Date(expirationDate),
   price: parseToBigDecimal(price),
   token: getSupportedTokenByName(rateId as SupportedTokenSymbol),
-  provider: providerId,
+  provider: { url: provider.url, provider: providerId },
```
Comment (jurajpiar) [inline src/api/rif-marketplace-cache/notifier/subscriptions/api.ts]:
```suggestion
  provider: { url, provider: providerId },
```

Comment (jurajpiar) [inline src/api/rif-marketplace-cache/notifier/subscriptions/api.ts]: could we get provider id in the provider object from the cache? seems silly having to map that.

File: src/components/organisms/notifier/provider/ProviderRegistrarForm.tsx [hunk 8]
```diff
@@ -27,11 +29,22 @@ const ProviderRegistrarForm: FC<ProviderRegistrarFormProps> = ({
...
     const notifierService: SubscriptionPlans = new SubscriptionPlans(url)
     notifierService.connect(reportError)
-    return notifierService.hasPlans()
+    providersApi.connect(reportError)
+    const providerService: ProvidersService = providersApi as ProvidersService
+    const urlIsUnregistered = await providerService.isUnregisteredURL(url.replace(/\/$/, ''))
```
Comment (jurajpiar) [inline src/components/organisms/notifier/provider/ProviderRegistrarForm.tsx]: make reusable (regex)

**Review by jurajpiar — APPROVED** [review]
just some details...
